File: logs/pollardsrho.go
package main

import (
    "fmt"
    "math/big"
	"math/rand"
	"time"
)

// Pollard's Rho iterator type
type PollardRhoIterator struct {
	point1 *Point
	point2 *Point
	X1 *Point
	X2 *Point
	a1 int64
	b1 int64
	a2 int64
	b2 int64
	X *Point
	a int64
	b int64
}

func NewPollardRhoIterator(P *Point, Q *Point) (*PollardRhoIterator) {

	// Create new iterator
	iter := new(PollardRhoIterator)
	iter.point1 = P
	iter.point2 = Q

	// Generate random pair a1, b1 within range of n
    s1 := rand.NewSource(time.Now().UnixNano())
    r1 := rand.New(s1)
    iter.a1 = r1.Int63n(myCurve.n)
    if (iter.a1 == 0) {
    	iter.a1++
    }
    s2 := rand.NewSource(time.Now().UnixNano())
    r2 := rand.New(s2)
    iter.b1 = r2.Int63n(myCurve.n)
    if (iter.b1 == 0) {
    	iter.b1++
    }

    // TBD
//    iter.a1 = 1
//    iter.b1 = 3
//    fmt.Printf("a1 = %d, b1 = %d\n", iter.a1, iter.b1)

    // Compute a1P + b1Q
    a1P := Multiply(iter.a1, P)
//    fmt.Printf("a1P = (%d, %d)\n\n\n", a1P.x, a1P.y)
    b1Q := Multiply(iter.b1, Q)
 //   fmt.Printf("b1Q = (%d, %d)\n\n\n", b1Q.x, b1Q.y)
    iter.X1 = Add(a1P, b1Q)
/*    if (iter.X1 == nil) {
    	fmt.Printf("X1 = nil\n\n\n")
    } else {
	    fmt.Printf("X1 = (%d, %d)\n\n\n", iter.X1.x, iter.X1.y)
	}
*/
    // Generate random pair a2, b2 within range of n
    s1 = rand.NewSource(time.Now().UnixNano())
    r1 = rand.New(s1)
    iter.a2 = r1.Int63n(myCurve.n)
    if (iter.a2 == 0) {
    	iter.a2++
    }
    s2 = rand.NewSource(time.Now().UnixNano())
    r2 = rand.New(s2)
    iter.b2 = r2.Int63n(myCurve.n)
    if (iter.b2 == 0) {
    	iter.b2++
    }

     // Compute a2P + b2Q
    a2P := Multiply(iter.a2, P)
//    fmt.Printf("a2P = (%d, %d)\n", a2P.x, a2P.y)
    b2Q := Multiply(iter.b2, Q)
//    fmt.Printf("b2Q = (%d, %d)\n", b2Q.x, b2Q.y)


    iter.X2 = Add(a2P, b2Q)
    // Initialize iterated values X, a, b 
    iter.X = nil
    iter.a = 0
    iter.b = 0

	return iter
}

func (iter *PollardRhoIterator) Next() (X *Point, a int64, b int64) {
	var i *big.Int

    // Partition the curve into three segments
    partitionSize := (myCurve.p / 3) + 1
    bigPartitionSize := big.NewInt(partitionSize)

    if (iter.X == nil) {
		i = big.NewInt(0)
    } else {
    	i = new(big.Int)
    	i.Div(iter.X.x, bigPartitionSize)
    }

    // Start with 0P, then add 1P to get 1P, where i == 0.
    // Next iteration is 2P (doubling 1P to get 2P), where i == 1.
    // Next iteration is 3P (adding P to 2P), where i == 2.
    if (i.Int64() == 0) {
    	iter.a += iter.a1
    	iter.b += iter.b1
//    	fmt.Printf("Iterating, i = 0, a = %d, b = %d\n", iter.a, iter.b)
    	iter.X = Add(iter.X, iter.X1)
    } else if (i.Int64() == 1) {
    	iter.a *= 2
    	iter.b *= 2
//    	fmt.Printf("Iterating, i = 1, a = %d, b = %d\n", iter.a, iter.b)
    	iter.X = Double(iter.X)
    } else if (i.Int64() == 2) {
    	iter.a += iter.a2
    	iter.b += iter.b2
//    	fmt.Printf("Iterating, i = 2, a = %d, b = %d\n", iter.a, iter.b)
    	iter.X = Add(iter.X, iter.X2)
    } else {
//    	fmt.Printf("Invalid i, returning.....\n")
    	return nil, 0, 0
    }

    // Take the a, b values mod n
    iter.a = mod(iter.a, myCurve.n)
    iter.b = mod(iter.b, myCurve.n)

    a = iter.a
    b = iter.b

//    fmt.Printf("a, b = %d, %d\n\n\n", a, b)

    return iter.X, a, b
}

// Pollard's Rho algorithm
func pollardRho(P *Point, Q *Point) (log int64, numSteps int64) {
	var i int64

	// Ensure points P and Q are both on the curve
	if ((isOnCurve(*P) == false) || (isOnCurve(*Q) == false)) {
		fmt.Printf("Exiting, point P or Q is not on curve!\n")
		return 0, numSteps
	}

	// Initialize numSteps to 0
	numSteps = 0

	tortoise := NewPollardRhoIterator(P, Q)
	if (tortoise == nil) {
		return 0, 0
	}

	// Copy values of tortoise to the hare iterator
	hare := *tortoise

	for i = 0; i < myCurve.n; i++ {
		X1, a1, b1 := tortoise.Next()

		// Hare skips over a step every time
		X2, a2, b2 := hare.Next()
		X2, a2, b2 = hare.Next()

		numSteps++

		// If (x1,y1) == (x2,y2), we've found a match (detected a cycle)
		if ((X1 == nil) && (X2 == nil)) {
			if (b1 == b2) {
				fmt.Printf("Generated random sequence divide by zero, retry\n")
				return 0, numSteps
			}
			log = (a1 - a2) * modInverse((b2 - b1), myCurve.n)
			log = mod(log, myCurve.n)
			return log, numSteps
		} else if ((X1 == nil) || (X2 == nil)) {
			numSteps++
		} else if ((X1.x.Cmp(X2.x) == 0) && (X1.y.Cmp(X2.y) == 0)) {
			if (b1 == b2) {
				fmt.Printf("Generated random sequence divide by zero, retry\n")
				return 0, numSteps
			}
			log = (a1 - a2) * modInverse((b2 - b1), myCurve.n)
			log = mod(log, myCurve.n)
			return log, numSteps
		} else {
			continue
		}
	}

	return 0, numSteps
}

File: logs/pollardsrho.py
#!/usr/bin/env python3

# This script makes use of another module: common.py, which can be
# found on GitHub:
#
#  https://github.com/andreacorbellini/ecc/blob/master/logs/common.py
#
# You must place that module on the same directory of this script
# prior to running it.

import random

from common import inverse_mod, tinycurve as curve


class PollardRhoSequence:

    def __init__(self, point1, point2):
        self.point1 = point1
        self.point2 = point2

        # Generate random pair (a1, b1)
        self.add_a1 = random.randrange(1, curve.n)
        self.add_b1 = random.randrange(1, curve.n)

#TBD
#        self.add_a1 = 3
#        self.add_b1 = 4

        # Compute a1P + b1Q
        print('')
        print('')
        print('')
        print('Generating new sequence....................')
        print('a1 = ', self.add_a1)
        print('b1 = ', self.add_b1)
        self.a1P = curve.mult(self.add_a1, point1)
        print('a1P = ', self.a1P)
        print('')
        print('')
        self.b1Q = curve.mult(self.add_b1, point2)
        print('b1Q = ', self.b1Q)
        print('')
        print('')
        self.add_x1 = curve.add(self.a1P, self.b1Q)

#        self.add_x1 = curve.add(
#            curve.mult(self.add_a1, point1),
#            curve.mult(self.add_b1, point2),
#        )

        print('X1 = ', self.add_x1)
        print('')
        print('')

        # Generate random pair (a2, b2)
        self.add_a2 = random.randrange(1, curve.n)
        self.add_b2 = random.randrange(1, curve.n)

#TBD
#        self.add_a2 = 3
#        self.add_b2 = 3

        # Compute a2P + b2Q
        print('a2 = ', self.add_a2)
        print('b2 = ', self.add_b2)
        self.a2P = curve.mult(self.add_a2, point1)
        self.b2Q = curve.mult(self.add_b2, point2)
        print('a2P = ', self.a2P)
        print('b2Q = ', self.b2Q)
        self.add_x2 = curve.add(self.a2P, self.b2Q)

#        self.add_x2 = curve.add(
#            curve.mult(self.add_a2, point1),
#            curve.mult(self.add_b2, point2),
#        )

        print('X2 = ', self.add_x2)
        print('')
        print('')

    def __iter__(self):
        # Partition the curve into 3 segments
        partition_size = curve.p // 3 + 1

        x = None
        a = 0
        b = 0

        # Start with 0P, then add 1P to get 1P, where i == 0.
        # Next iteration is 2P (doubling 1P to get 2P), where i == 1.
        # Next iteration is 3P (adding P to 2P), where i == 2.
        while True:
            if x is None:
                i = 0
                print('Partition size = ', partition_size)
            else:
                i = x[0] // partition_size
                print('Partition size = ', partition_size)

            if i == 0:
                # x is either the point at infinity (None), or is in the first
                # third of the plane (x[0] <= curve.p / 3).
                a += self.add_a1
                b += self.add_b1
                print('Iterating, i = 0, a =', a, ', b =', b)
                x = curve.add(x, self.add_x1)
            elif i == 1:
                # x is in the second third of the plane
                # (curve.p / 3 < x[0] <= curve.p * 2 / 3).
                a *= 2
                b *= 2
                print('Iterating, i = 1, a =', a, ', b =', b)
                x = curve.double(x)
            elif i == 2:
                # x is in the last third of the plane (x[0] > curve.p * 2 / 3).
                a += self.add_a2
                b += self.add_b2
                print('Iterating, i = 2, a =', a, ', b =', b)
                x = curve.add(x, self.add_x2)
            else:
                raise AssertionError(i)

            print('x =', x)
            a = a % curve.n
            b = b % curve.n
            print('a, b = ', a, b)
            print('')
            print('')
            yield x, a, b


def log(p, q, counter=None):
    assert curve.is_on_curve(p)
    assert curve.is_on_curve(q)

    # Pollard's Rho may fail sometimes: it may find a1 == a2 and b1 == b2,
    # leading to a division by zero error. Because PollardRhoSequence uses
    # random coefficients, we have more chances of finding the logarithm
    # if we try again, without affecting the asymptotic time complexity.
    # We try at most three times before giving up.
    for i in range(3):
        sequence = PollardRhoSequence(p, q)

        tortoise = iter(sequence)
        hare = iter(sequence)

        print('Iterate over the sequences...................')
        # The range is from 0 to curve.n - 1, but actually the algorithm will
        # stop much sooner (either finding the logarithm, or failing with a
        # division by zero).
        for j in range(curve.n):
            print('One tortoise step:')
            x1, a1, b1 = next(tortoise)

            # Hare skips over a step
            print('Two hare steps')
            x2, a2, b2 = next(hare)
            x2, a2, b2 = next(hare)

            print('a1 =', a1, ', b1 =', b1, ', a2 =', a2, ' b2 =', b2)

            # We have found a match (detected a cycle) if x1 == x2
            if x1 == x2:
                if b1 == b2:
                    # This would lead to a division by zero. Try with
                    # another random sequence.
                    break

                x = (a1 - a2) * inverse_mod(b2 - b1, curve.n)
                logarithm = x % curve.n
                steps = i * curve.n + j + 1
                return logarithm, steps

    raise AssertionError('logarithm not found')


def main():
    x = random.randrange(1, curve.n)
    p = curve.g
    q = curve.mult(x, p)

    print('Curve: {}'.format(curve))
    print('Curve order: {}'.format(curve.n))
    print('p = (0x{:x}, 0x{:x})'.format(*p))
    print('q = (0x{:x}, 0x{:x})'.format(*q))
    print(x, '* p = q')

    y, steps = log(p, q)
    print('log(p, q) =', y)
    print('Took', steps, 'steps')

    assert x == y


if __name__ == '__main__':
    main()
